src/datajson.c

```diff
@@ -0,0 +1,1393 @@
+    return 0;
+}
+
+Dataset *DatajsonGet(const char *name, enum DatasetTypes type, const char *load, uint64_t memcap,
+        uint32_t hashsize, char *json_key_value, char *json_array_key)
+{
+    uint64_t default_memcap = 0;
+    uint32_t default_hashsize = 0;
+    if (strlen(name) > DATASET_NAME_MAX_LEN) {
+        return NULL;
+    }
+
+    DatasetLock();
+    Dataset *set = DatasetSearchByName(name);
+    if (set) {
+        if (type != DATASET_TYPE_NOTSET && set->type != type) {
+            SCLogError("dataset %s already "
+                       "exists and is of type %u",
+                    set->name, set->type);
+            DatasetUnlock();
+            return NULL;
+        }
+
+        if (load == NULL || strlen(load) == 0) {
+            // OK, rule keyword doesn't have to set state/load,
+            // even when yaml set has set it.
+        } else {
+            if ((load == NULL && strlen(set->load) > 0) ||
+                    (load != NULL && strcmp(set->load, load) != 0)) {
+                SCLogError("dataset %s load mismatch: %s != %s", set->name, set->load, load);
+                DatasetUnlock();
+                return NULL;
+            }
+        }
+
+        DatasetUnlock();
+        return set;
+    }
+
+    if (type == DATASET_TYPE_NOTSET) {
+        SCLogError("dataset %s not defined", name);
+        goto out_err;
+    }
+
+    set = DatasetAlloc(name);
+    if (set == NULL) {
+        goto out_err;
+    }
+
+    strlcpy(set->name, name, sizeof(set->name));
+    set->type = type;
+    if (load && strlen(load)) {
+        strlcpy(set->load, load, sizeof(set->load));
+        SCLogDebug("set \'%s\' loading \'%s\' from \'%s\'", set->name, load, set->load);
+    }
+
+    static const char conf_format_str[] = "datasets.%s.hash";
+    char cnf_name[DATASET_NAME_MAX_LEN + (sizeof(conf_format_str) / sizeof(char))];
+    int p_ret = snprintf(cnf_name, sizeof(cnf_name), conf_format_str, name);
+    if (p_ret == 0) {
+        SCLogError("Can't build configuration variable for set: '%s'", name);
+        goto out_err;
+    }
+
+    DatasetGetDefaultMemcap(&default_memcap, &default_hashsize);
+    switch (type) {
+        case DATASET_TYPE_MD5:
+            set->hash = THashInit(cnf_name, sizeof(Md5TypeJson), Md5StrJsonSet, Md5StrJsonFree,
+                    Md5StrJsonHash, Md5StrJsonCompare, NULL, NULL, load != NULL ? 1 : 0,
+                    memcap > 0 ? memcap : default_memcap,
+                    hashsize > 0 ? hashsize : default_hashsize);
+            if (set->hash == NULL)
+                goto out_err;
+            if (DatajsonLoadMd5(set, json_key_value, json_array_key) < 0)
+                goto out_err;
+            break;
+        case DATASET_TYPE_STRING:
+            set->hash = THashInit(cnf_name, sizeof(StringTypeJson), StringJsonSet, StringJsonFree,
+                    StringJsonHash, StringJsonCompare, NULL, NULL, load != NULL ? 1 : 0,
```

Comment:
It seems like we might be missing a length getter here which is used to calculate memuse in THash API. This is needed for the string type as we do not have a definite size for it. wdyt?
See: https://redmine.openinfosecfoundation.org/issues/3910